src/main/java/org/spongepowered/api/text/channel/impl/EventPostingMessageChannel.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.channel.impl;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.event.SpongeEventFactory;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.cause.NamedCause;
+import org.spongepowered.api.event.message.MessageChannelEvent;
+import org.spongepowered.api.event.message.MessageEvent;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.channel.MessageChannel;
+import org.spongepowered.api.text.chat.ChatType;
+
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link MessageChannel} which posts {@link MessageChannelEvent} before sending a message.
+ */
+public class EventPostingMessageChannel extends DelegateMessageChannel {
+
+    public EventPostingMessageChannel(final MessageChannel delegate) {
+        super(delegate);
+    }
+
+    @Override
+    public void send(@Nullable final Object sender, final Text original, final ChatType type) {
+        final MessageChannelEvent event = SpongeEventFactory.createMessageChannelEvent(
+                Cause.of(NamedCause.source(sender), NamedCause.of("Type", type)),
```

Comment:
You could handle the send method without firing the event if the sender is null, or you could just `return` out of the method without sending the text. I think calling `super.send(...)` is best if the sender is null just to avoid having messages disappear.